Geomap: Document the Geomap panel (#37221)
grafana/grafana: docs/sources/panels/visualizations/geomap.md

# Geomap

{{< figure src="/static/img/docs/geomap-panel/geomap_example.png" max-width="1200px" caption="Geomap panel" >}}

> The Geomap panel visualization allows you to view and customize the world map using geospatial data in respect to geographical locations. Various overlay styles and map view settings can be configured to easily focus on the important location-based characteristics of the data.

Reviewer: Suggestion: `The Geomap panel visualization allows you to view and customize the world map using geospatial data.` (Delete `in respect to geographical locations` - might be redundant.)

## Base layer

> Base layer loads in a blank world map from the tile server to the Grafana panel. There are several base layer options available for you to choose from, each with specific configuration options to style the base map. The default base layer is CartoDB base map. Custom default base layers can be defined using the `.ini` configuration file.

Reviewer 2: ```suggestion
The base layer loads in a blank world map from the tile server to the Grafana panel. Several base layer options are available each with specific configuration options to style the base map. The default base layer is CartoDB base map. Custom default base layers can be defined in the `.ini` configuration file.
```

> Base layer loads in a blank world map from the tile server to the Grafana panel. There are several base layer options available for you to choose from, each with their specific configuration options to style the base map. Default base layer is CartoDB base map, but can be specified differently through the `.ini` configuration file.

Reviewer: Suggestion: `Base layer loads in a blank world map from the tile server to the Grafana panel. There are several base layer options available for you to choose from, each with specific configuration options to style the base map. The default base layer is CartoDB base map. Custom default base layers can be defined using the `.ini` configuration file.

![Base layer options](/static/img/docs/geomap-panel/geomap_base_layer.png)

### Configure the default base layer with provisioning

> It is possible to configure the default base map using config files with Grafana’s provisioning system. You can read more about how it works and all the settings you can set on the [provisioning docs page]({{< relref "../administration/provisioning.md" >}}).

Reviewer 2: ```suggestion
You can configure the default base map using config files with Grafana’s provisioning system. For more information on all the settings, refer to the [provisioning docs page]({{< relref "../administration/provisioning.md" >}}).
```

> `default_baselayer_config` is the JSON configuration used to define the default base map. There are currently four base map options to choose from: `carto`, `esri-xyz`, `osm-standard`, `xyz`. Here are some provisioning examples for each base map options.

Reviewer 2: ```suggestion
Use the JSON configuration option `default_baselayer_config` to define the default base map. There are currently four base map options to choose from: `carto`, `esri-xyz`, `osm-standard`, `xyz`. Here are some provisioning examples for each base map option.
```

> - **carto** loads the CartoDB tile server. There is a choice between `auto`, `dark`, and `light` theme for the base map and can be set as shown below. `showLabels` tag determines whether or not to show the Country details on top of the map.

Reviewer 2: ```suggestion
- **carto** loads the CartoDB tile server. You can choose from `auto`, `dark`, and `light` theme for the base map and can be set as shown below.  The `showLabels` tag determines whether or not Grafana shows the Country details on top of the map. Here is an example:
```

> - **carto** loads the CartoDB tile server. There is a choice between `auto`, `dark`, and `light` theme for the base map and can be set as shown below. `showLabels` tag determines whether or not to show the Country details on top of the map.

Reviewer 2: carto can be a h3. However, I am ok with this formatting as well.

> - **carto** loads the CartoDB tile server. There is a choice between `auto`, `dark`, and `light` theme for the base map and can be set correspondingly as shown below. `showLabels` tag determines whether or not to show the Country details on top of the map.

Reviewer: Suggestion: remove `correspondingly`

Reviewer 3: Thank you! Addressed your comments :D

```ini
geomap_default_baselayer = `{
  "type": "carto",
  "config": {
    "theme": "auto",
    "showLabels": true
  }
}`
```

> - **esri-xyz** loads the ESRI tile server. There are already multiple server instances implemented to show the various map styles: `world-imagery`, `world-physical`, `topo`, `usa-topo`, and `ocean`. There is also a `custom` server option which allows to configure your own ArcGIS map server.

Reviewer 2: ```suggestion
- **esri-xyz** loads the ESRI tile server. There are already multiple server instances implemented to show the various map styles: `world-imagery`, `world-physical`, `topo`, `usa-topo`, and `ocean`. The `custom` server option allows you to configure your own ArcGIS map server. Here are some examples:
```

```ini
geomap_default_baselayer = `{
  "type": "esri-xyz",
  "config": {
    "server": "world-imagery"
  }
}`
```

```ini
geomap_default_baselayer = `{   
  "type": "esri-xyz",
  "config": {
    "server": "custom",
    "url": "[tile server url]",
    "attribution": "[tile server attribution]"
  }
}`
```

> - **osm-standard** loads the OpenStreetMap tile server. There is no additional `config` fields needed to be set and can be left blank. 

Reviewer 2: ```suggestion
- **osm-standard** loads the OpenStreetMap tile server. There are no additional configurations needed and the `config` fields can be left blank. 
```

```ini
default_baselayer_config = `{
  "type": "osm-standard",
  "config": {}
}`
```

> - **xyz** loads a custom tile server defined by the user.  A valid tile server `url`, with {z}/{x}/{y}, has to be set in order for this option to properly load a default base map.

Reviewer 2: ```suggestion
- **xyz** loads a custom tile server defined by the user.  Set a valid tile server `url`, with {z}/{x}/{y} for this option in order to properly load a default base map.
```

```ini
default_baselayer_config = `{
  "type": "xyz",
  "config": {
    "attribution": "Open street map",
    "url": "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
  }
}`
```

> `enable_custom_baselayers` gives the option to enable/disable custom open source base maps that are already implemented. This will be defaulted to `true`.

Reviewer 2: ```suggestion
`enable_custom_baselayers` allows you to enable or disable custom open source base maps that are already implemented. The default is `true`.
```

## Data layer

> Data layer in the Geomap plugin determines how you visualize geospatial data on top of the base map.

Reviewer 2: ```suggestion
The data layer in the Geomap plugin determines how you visualize geospatial data on top of the base map.
```

### Location

> The Geomap panel needs a source of geographical data. The data comes from a database query, and there are four mapping options for your data.

Reviewer 2: ```suggestion
The Geomap panel needs a source of geographical data. This data comes from a database query, and there are four mapping options for your data.
```

> - **Auto** will automatically search for location data. Use this data option when your query is based on one of the following names for data fields.

Reviewer 2: ```suggestion
- **Auto** automatically searches for location data. Use this option when your query is based on one of the following names for data fields.
```

  - geohash: “geohash”
  - latitude: “latitude”, “lat”
  - longitude: “longitude”, “lng”, “lon”
  - lookup: “lookup”

> - **Coords** specifies that your query holds coordinate data. You will be prompted to select numeric data fields for latitude and longitude from your database query.

Reviewer 2: ```suggestion
- **Coords** specifies that your query holds coordinate data. You will get prompted to select numeric data fields for latitude and longitude from your database query.
```

> - **Geohash** specifies that your query holds geohash data. You will be prompted to select a string data field for geohash from your database query.

Reviewer 2: ```suggestion
- **Geohash** specifies that your query holds geohash data. You will get prompted to select a string data field for the geohash from your database query.
```

> - **Lookup** specifies that your query holds location name data that needs to be mapped to a value. You will be prompted to select the lookup field from your database query and a gazetteer. The gazeteer is the directory that will be used to map your queried data to a geographical point.

Reviewer 2: ```suggestion
- **Lookup** specifies that your query holds location name data that needs to be mapped to a value. You will get prompted to select the lookup field from your database query and a gazetteer. The gazetteer is the directory that is used to map your queried data to a geographical point.
```

### Markers Layer

> ![Markers Layer](/static/img/docs/geomap-panel/geomap_markers_layer.png)

Reviewer 2: Move image after inttro. Not a good practice to open a section with an image.

> Markers layer allows displaying data points as different marker shapes such as circle, square, triangle, star, and more.

Reviewer 2: ```suggestion
The markers layer allows you to display data points as different marker shapes such as circle, squares, triangles, stars, and more.
```

![Markers Layer Options](/static/img/docs/geomap-panel/geomap_markers_options.png)

- **Marker Color** configures the color of the marker. The default `Fixed size` keeps all points a single color. There is an alternate option to have multiple colors depending on the data point values and the threshold set at the `Thresholds` section.
- **Marker Size** configures the size of the marker. Default is `Fixed size`, making all marker size the same regardless of the data points. However, there is also an option to scale the circles to the corresponding data points. `Min` and `Max` marker size has to be set such that the Marker layer can scale within these range.

> - **Marker Shape** gives flexibility to visualize the data points differently. 

Reviewer 2: ```suggestion
- **Marker Shape** provides you with the flexibility to visualize the data points differently. 
```

  - Circle
  - Square
  - Triangle
  - Star
  - Cross
  - X
- **Fill Opacity** configures the transparency of each marker.

### Heatmap Layer

> ![Heatmap Layer](/static/img/docs/geomap-panel/geomap_heatmap_layer.png)

Reviewer 2: Move image.

Reviewer 3: Thank you! Addressed the comments. 

Regarding images in documentations, where should I be adding those images / where is the Grafana/website repo? https://github.com/grafana/grafana/tree/main/docs

Reviewer 2: In the website repo: https://github.com/grafana/website/tree/master/static/static/img/docs, make an folder geomap and add images there. 

Here is useful information on how to name images: https://github.com/grafana/grafana/blob/main/contribute/style-guides/documentation-style-guide.md#images. I encourage users to add a suffix for the version when the image was taken: "image-8-1-0.png".

Reviewer 3: Is this a private repo? I get a page not found with https://github.com/grafana/website/tree/master/static/static/img/docs

Reviewer 2: Not an external repo. If you want, share the images and I can upload them.

Reviewer 4: Hi @achatterjee-grafana We will upload the images here. But why is the docs repo not public? Are the docs generated? This is very confusing for non-Grafana contributors to the project.

Reviewer 2: Tagging @ryantxu, so he can explain why the website repo is not accessible. Ryan, the issue came up as KIm needs to upload images. Thanks.

Reviewer 3: > Not an external repo. If you want, share the images and I can upload them.

geomap-example-8-1-0.png
![geomap-example-8-1-0](https://user-images.githubusercontent.com/43913498/127058989-bd708f3a-60a8-4404-b00e-160f8cd522d7.png)

geomap-baselayer-8-1-0.png
![geomap-baselayer-8-1-0](https://user-images.githubusercontent.com/43913498/127059365-3b9a639c-bf22-4eb9-9d9b-feefe16eb0e6.png)

geomap-markers-8-1-0.png
![geomap-markers-8-1-0](https://user-images.githubusercontent.com/43913498/127060400-4a399558-51c6-4eb4-afbc-dc0b6e63ef6b.png)

geomap-markers-options-8-1-0.png
<img width="411" alt="geomap-markers-options-8-1-0" src="https://user-images.githubusercontent.com/43913498/127059955-d0974909-4acd-4d2b-bf2c-cc189b93dc09.png">

geomap-heatmap-8-1-0.png
![geomap-heatmap-8-1-0](https://user-images.githubusercontent.com/43913498/127060120-f7f08ebb-507e-4278-a1a0-f107b51f47b3.png)

geomap-heatmap-options-8-1-0.png
<img width="409" alt="geomap-heatmap-options-8-1-0" src="https://user-images.githubusercontent.com/43913498/127060185-d2078ff1-db2d-464d-8d96-a0ce2cb7893b.png">

> Heatmap layer cluster data points to visualize locations with different densities. To add a heatmap layer, click on the drop down menu under Data Layer and choose `Heatmap`. Similar to `Markers`, you will be prompted with various options determine which data points to visualize and how. 

Reviewer 2: ```suggestion
The heatmap layer clusters various data points to visualize locations with different densities. 

To add a heatmap layer:
Click on the drop down menu under Data Layer and choose `Heatmap`. 

Similar to `Markers`, you are prompted with various options to determine which data points to visualize and how. 
```

- **Lookup** specifies that your query holds location name data that needs to be mapped to a value. You will get prompted to select the lookup field from your database query and a gazetteer. The gazetteer is the directory that is used to map your queried data to a geographical point.

### Markers layer

The markers layer allows you to display data points as different marker shapes such as circle, squares, triangles, stars, and more.

![Markers Layer](/static/img/docs/geomap-panel/geomap-markers-8-1-0.png)

![Markers Layer Options](/static/img/docs/geomap-panel/geomap-markers-options-8-1-0.png)

- **Marker color** configures the color of the marker. The default `Fixed size` keeps all points a single color. There is an alternate option to have multiple colors depending on the data point values and the threshold set at the `Thresholds` section.
- **Marker size** configures the size of the marker. Default is `Fixed size`, making all marker size the same regardless of the data points. However, there is also an option to scale the circles to the corresponding data points. `Min` and `Max` marker size has to be set such that the Marker layer can scale within these range.

> - **Marker shape** provides you with the flexibility to visualize the data points differently. 

Reviewer 2: What does the UI say? We should stay true to the UI for labels. Thanks :)

Reviewer 3: Makes sense, those are changed now. I wasn't sure if I should use sentence case for these as well. :)

Reviewer 2: No worries! Many thanks for your contribution!! I will upload the images, then do a final review and merge.
